File: python_version/linked-lists/insertion-sort-list-147.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def insertionSortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(0)
        dummy.next = head
        nodeToInsert = head
        while head and head.next:
          if head.val > head.next.val:
            nodeToInsert = head.next
            nodeToInsertPre = dummy
            while nodeToInsertPre.next.val < nodeToInsert.val:
              nodeToInsertPre = nodeToInsertPre.next
            head.next = nodeToInsert.next
            nodeToInsert.next = nodeToInsertPre.next
            nodeToInsertPre.next = nodeToInsert
          else:
            head = head.next
        return dummy.next

File: src/linked-lists/insertion-sort-list-147.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode insertionSortList(ListNode head) {
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode nodeToInsert, nodePrev= head;
        
        while(head != null && head.next != null){
            if(head.val > head.next.val){
                nodeToInsert = head.next;
                nodePrev = dummy;
                //boundary: first node n.next.val >= nodeToInsert.val
                while(nodePrev.next.val < nodeToInsert.val){
                    nodePrev = nodePrev.next;
                }
                // then we could make an insertion
                head.next = nodeToInsert.next;
                nodeToInsert.next = nodePrev.next;
                nodePrev.next = nodeToInsert;
            }
            else{
                head = head.next;
            }
        }
        
        return dummy.next;
    }
}